src/recycle/shared_pool.hpp

```diff
@@ -267,6 +285,34 @@ class shared_pool
 
             lock_type lock(m_mutex);
             m_free_list.push_back(resource);
+            m_size--;
+        }
+
+        constexpr std::size_t capacity () noexcept
+        {
+            lock_type lock(m_mutex);
+            return m_max_capacity;
+        }
+
+        constexpr void reserve (std::size_t new_cap)
+        {
+            std::list<value_ptr> pending_resources;
+            {
+                lock_type lock(m_mutex);
+                if  (new_cap <= m_max_capacity)
+                    return;
+
+                auto diff = new_cap - m_size;
+                for (auto i = 0; i < diff; i++) {
+                    value_ptr resource = m_allocate();
+                    auto pool = impl::shared_from_this();
+
+                    pending_resources.emplace_back(resource.get(), deleter(pool, resource));
+                }
+
+                m_max_capacity = new_cap;
+            }
```

Comment:
It would have a critical section here. The return of capacity() and unused_free() won't match.
Besides, while we recycle buffer here, the second thread which calls the pool->allocate() may break the atomic.
Should I use the thread-local lock here, we need make a rule which thread calls reserve(), which thread can allocate().